packages/common/src/EntityPin.ts

```diff
@@ -0,0 +1,58 @@
+import { Entity } from "./Entity";
+import { d3SelectionType } from "./Widget";
+
+export class EntityPin extends Entity {
+    protected _element_textbox: d3SelectionType;
+    constructor() {
+        super();
+    }
+    enter(domNode, element) {
+        super.enter(domNode, element);
+    }
+    update() {
+        super.update.apply(this, arguments);
+
+        this._icon_widget.render();
+        this._desc_widget.render();
+        this._title_widget.render();
+        const title_bbox = this._title_widget.getBBox(true);
+        const icon_bbox = this._icon_widget.getBBox(true);
+        const desc_bbox = this._desc_widget.getBBox(true);
+        const _anno_h = this.annotationIcons().length > 0 ? this.annotationDiameter() + this.padding() : 0;
+        const background_bbox = {
+            width: title_bbox.width + (this.padding() * 2),
+            height: icon_bbox.height + title_bbox.height + _anno_h + this.arrowHeight() + desc_bbox.height + (this.padding() * 5)
+        };
+
+        this._background_widget
+            .height(background_bbox.height)
+            .width(background_bbox.width);
+        const _icon_y = -background_bbox.height + (icon_bbox.height / 2) + this.padding();
+        const _title_y = _icon_y + (icon_bbox.height / 2) + (title_bbox.height / 2) + this.padding();
+        const _desc_y = _title_y + (title_bbox.height / 2) + (desc_bbox.height / 2) + this.padding();
+        const _anno_y = _desc_y + (desc_bbox.height / 2) + this.padding();
+        this._title_widget.move({
+            x: 0,
+            y: _title_y
+        });
+        this._desc_widget.move({
+            x: 0,
+            y: _desc_y
+        });
+        this._icon_widget.move({
+            x: 0,
+            y: _icon_y
+        });
+        let anno_w_sum = 0;
+        const anno_gap = this.annotationSpacing();
+        for (const i in this._annotation_widgets) {
+            const anno_bbox = this._annotation_widgets[i]._bbox;
+            this._annotation_widgets[i].move({
+                x: -anno_w_sum + (background_bbox.width / 2) - anno_bbox.width,
+                y: _anno_y + (anno_bbox.height / 2)
+            });
+            anno_w_sum += anno_bbox.width + anno_gap;
+        }
```

Comment:
You should have a common function for this which positions the row of annotations at a given place.